File: 1395-minimum-time-visiting-all-points/minimum-time-visiting-all-points.py
class Solution(object):
    def minTimeToVisitAllPoints(self, points):
        """
        Algorithm:
        1. Start from the last point in the list(x1, y1).
        2. Repeatedly move to the next point(x2, y2).
        3. For each move from (x1, y1) to (x2, y2):
           - The minimum time required is: max(|x2 - x1|, |y2 - y1|)
        4. Accumulate and Return the total time.
        """
        # Appraoch 1: modifying points
        # min_time = 0
        # x1, y1 = points.pop()
        # while points:
        #     x2, y2 = points.pop()
        #     min_time += max(abs(x2-x1),abs(y2-y1))
        #     x1, y1 = x2, y2
        # return min_time

        # Appraoch 2: without modifying points
        min_time = 0
        n = len(points)
        for i in range(1,n):
            x1, y1 = points[i-1]
            x2, y2 = points[i]
            min_time += max(abs(x2-x1),abs(y2-y1))
        return min_time

"""
Complexity:
Time : O(n) - Traversal through points
Space : O(1) - Constant
"""

File: 1395-minimum-time-visiting-all-points/minimum-time-visiting-all-points.cpp
class Solution {
public:
    int minTimeToVisitAllPoints(vector<vector<int>>& points) {
        int min_cost = 0; // Initial time
        for (int i = 0; i < points.size() - 1; i++){
            int curr_x = points[i][0];
            int curr_y = points[i][1];
            int targ_x = points[i+1][0];
            int targ_y = points[i+1][1];

            min_cost += max(abs(targ_x - curr_x), abs(targ_y - curr_y));
        }
        return min_cost;
    }
};